opensrp-app/src/test/java/org/smartregister/sync/intent/SyncIntentServiceTest.java

```diff
@@ -0,0 +1,86 @@
+package org.smartregister.sync.intent;
+
+import android.content.Context;
+import android.content.Intent;
+import android.content.pm.PackageManager;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.powermock.reflect.Whitebox;
+import org.robolectric.RuntimeEnvironment;
+import org.smartregister.BaseRobolectricUnitTest;
+import org.smartregister.CoreLibrary;
+import org.smartregister.SyncConfiguration;
+import org.smartregister.domain.FetchStatus;
+import org.smartregister.receiver.SyncStatusBroadcastReceiver;
+import org.smartregister.util.SyncUtils;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * Created by Richard Kareko on 7/21/20.
+ */
+
+public class SyncIntentServiceTest extends BaseRobolectricUnitTest {
+
+    @Mock
+    private SyncUtils syncUtils;
+
+    @Mock
+    SyncConfiguration syncConfiguration;
+
+    @Captor
+    private ArgumentCaptor<Intent>  intentArgumentCaptor;
+
+    private Context context = RuntimeEnvironment.application;
+
+    private SyncIntentService syncIntentService;
+
+    @Before
+    public void setUp() {
+        Whitebox.setInternalState(CoreLibrary.getInstance(), "syncConfiguration", syncConfiguration);
+        MockitoAnnotations.initMocks(this);
+        syncIntentService = new SyncIntentService();
+        syncIntentService.init(context);
+        Whitebox.setInternalState(syncIntentService, "mBase", RuntimeEnvironment.application);
+    }
+
+    @Test
+    public void testInit() {
+        assertNotNull(Whitebox.getInternalState(syncIntentService, "httpAgent"));
+        assertNotNull(Whitebox.getInternalState(syncIntentService, "syncUtils"));
+        assertNotNull(Whitebox.getInternalState(syncIntentService, "context"));
+    }
+
+    @Test
+    public void testHandleSyncSendFetchStartedBroadCast() throws PackageManager.NameNotFoundException {
+        syncIntentService = spy(syncIntentService);
+        Whitebox.setInternalState(syncIntentService, "syncUtils", syncUtils);
+        when(syncUtils.verifyAuthorization()).thenReturn(true);
+        when(syncUtils.isAppVersionAllowed()).thenReturn(false);
```

Comment:
This is interesting, fetch really shouldn't happen if this is false.

So not sure why this test passes.

Reply:
The check for `isAppVersionAllowed` id done later in the code after the broadcast has been sent as seen [here](https://coveralls.io/builds/32342498/source?filename=opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java#L98)

Reply:
I guess the point was that the message should not be broadcast until we are sure fetch is actually happening.

Which atm only happens [here](https://github.com/OpenSRP/opensrp-client-core/blob/9611aa1ccb9308c3406b7fdaba8ab9bc9cc79cb9/opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java#L105).

Reply:
Fetch does not happen for the other `if` branches.

Reply:
@ndegwamartin @githengi Would like to hear your thoughts on this. Should the broadcast message be moved from [here](https://coveralls.io/builds/32342498/source?filename=opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java#L78) to [here](https://github.com/OpenSRP/opensrp-client-core/blob/9611aa1ccb9308c3406b7fdaba8ab9bc9cc79cb9/opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java#L114)

Reply:
Hmm, maybe not unless you do some more refactoring. The sync process is comprised of two items, `pushToServer` and `pullECFromServer`. Push does not fall under the _minAppVersionAllowed_ condition since it is triggered before then

Reply:
I think we should keep `minAppVersionAllowed` separate from `pushToServer` as currently is. Data on device should still be allowed to be uploaded even if app is older and deprecated.

Reply:
The `FetchStatus.fetchStarted` is a bit of a misnomer since it signals the beginning of sync (both push and pull) though it should ideally apply to just pull (it's synonymous with fetch)?

So the question is whether to broadcast that only when an actual pull is happening, which doesn't happen for the first two branches.

Reply:
I agree that `FetchStatus.fetchStarted` is a misnomer but it is used to signal that the sync process has started as such moving the broadcast to when the pull is happening would in essence mean that we are not considering the push to server bit as part of sync.